### ISSUE TITLE
Add support for using lock files

### ISSUE THREAD

**Issue by leighmcculloch** [obi1kenobi/cargo-semver-checks#991]
### Describe your use case

When running `cargo semver-checks` no lock files appear to be used, either for the current / local version of the crate, or for versions of the crate downloaded from crates.io.

This creates a reproducibility issue. When crates publish releases that do not follow semver exactly, semver may fail to compile if no lock files is used. It also creates a situation where one run of the semver-check and a subsequent run can have different inputs and outputs, meaning no reproducibility in CI.

This creates a security issue. Some times use lock files to ensure that the developer is exclusively in control of what versions of dependencies are installed and in use. This is to avoid the problem that just before running a command a crate is compromised and a malware version is uploaded for it. Today using semver-checks as it is appears to expose us to that possibility when other cargo commands provide the --locked option.

### Describe the solution you'd like

The ability to pass `--locked` to the semver-checks command, and for the lock file to be used.

**Comment by obi1kenobi** [comment]
Hi, thanks for opening this issue.

I'd like to get a bit more context on the workflow you're using and the specific use case, to make sure I understand everything properly.

Additionally, could you say more about this part?

> Some times use lock files to ensure that the developer is exclusively in control of what versions of dependencies are installed and in use.

I'm having trouble reconciling this with the fact that developers specify the ranges or exact versions they want in `Cargo.toml`. The lockfile would only contain a version that is explicitly allowed by the range or exact bound set in `Cargo.toml`.

Separately, could you describe the possible security issue you are referring to?

Note that `--locked` exists for `cargo install` which is specific to binary crates. Library crates, which is what `cargo-semver-checks` analyzes, do not always contain a `Cargo.lock` file as part of their package index entry. There has been some motion toward changing that in the future, but unfortunately we can't add those files retroactively: https://github.com/rust-lang/cargo/issues/13447

**Comment by leighmcculloch** [comment]
Hi @obi1kenobi! Thanks for asking more questions.

We use `cargo-semver-checks` in the following workflow:
- https://github.com/stellar/rs-soroban-sdk/blob/fc32690edd223fd01054c625d9c6dca01bd99e67/.github/workflows/rust.yml#L55-L64

The `soroban-sdk` crate in the repos is dependent on the `arbitrary` crate with version specifier `1.3.0`, so all versions `>= 1.3.0, < 2.0.0` are supported. Which I think is reasonable assuming the crate follows semver.

`arbitrary` `1.4.0` was released that unfortunately contains some breaking changes (https://github.com/rust-fuzz/arbitrary/issues/203). 

When `cargo-semver-checks` runs now, comparing the `main` branch against the last release 21.7.6, it fails to build because ignoring the lock file it pulls in arbitrary 1.4.0.

Even if I update the dependency version specifier on the `main` branch to `~1.3.0`, so compatible versions are constrained to `>= 1.3.0, < 1.4.0`, that will only apply to the main branch and not the previous release being compared to. 21.7.6 will fail to build because it was dependent on `1.3.0` which says it is compatible with `1.4.0`.

https://github.com/stellar/rs-soroban-sdk/actions/runs/11672542695/job/32501403700

> Some times use lock files to ensure that the developer is exclusively in control of what versions of dependencies are installed and in use.

We commit lock files so that our builds are consistent, and we're 100% in control of what dependencies, and what versions of dependencies, are in use in our CI environments (that may have access to credentials like a crates.io auth token), and in use on our workstations. We choose when we change dependencies in our development workflows.

**Comment by obi1kenobi** [comment]
Ah hmm interesting. Tricky case for sure, thank you for bringing it to my attention. Thanks also for the additional context.

In addition to worrying about https://github.com/rust-lang/cargo/issues/13447, I need to figure out a couple more things here:

When crates re-export foreign crates' types, ignoring the lockfile is _usually_ the desirable behavior because it accurately reflects what would happen on a new `cargo add`. Otherwise it's possible for a project's `cargo-semver-checks` run to succeed based on the lockfiles but to be broken for downstream projects on `cargo add` due to a dependency. For example, if your crate had been re-exporting things form `arbitrary`, that case would have happened right here.

We also cannot say "use the lockfile in the current project but if it's missing on crates.io just re-lock from there." That would lead to "back to the future" style bugs: say your library re-exports a 3rd party crate's type, then re-locking the crates.io version can produce a newer version of that dependency in the baseline than in the current version. API _additions_ to that newer dependency version (in the baseline) are not present in the older dependency version (in your current lockfile) — it looks like your current package has _deleted_ things from its public API! This has happened before e.g. in #167.

A good solution to this issue shouldn't be "just a different kind of footgun" than we already have. I need to figure out how to make that happen, and whether https://github.com/rust-lang/cargo/issues/13447 makes progress or not is also a key question here.

Sorry I don't have a more immediately-useful answer here!

Btw, if cargo-semver-checks has been useful to your project, I'd really appreciate if your org started sponsoring me on GitHub 💖

**Comment by leighmcculloch** [comment]
Similar tensions exist with the use of `--locked` on the `cargo install` command. On that command for reproducibility one often wants to lock, but when there are security fixes in dependencies and versions have been yanked it's often desirable to `cargo install` without `--locked` to get that one dependency upgraded and there's no way to easily install with that one dependency upgraded. It's not quite the same thing, but the presence of `--locked` at least allows the option to test with a dependency graph set in time to something.

It may be desirable for some folks to run `cargo semver-checks` with and without `--locked` so as to get a view on how things have changed with what is likely (but not guaranteed) to be the minimum supported versions, and the latest bleeding edge versions.

Perfect may be the enemy of good here, but I understand the path forward is pretty unclear.

**Comment by leighmcculloch** [comment]
Based on the options on the current command, I think it would be reasonable to have `--locked` which applied to only the current, not baseline, since that's what most options already do on the command. Then have `--baseline-locked` that applies to the baseline, following the existing pattern.

In either case if there's no lock file, then I think it's reasonable to fail. The tool doesn't need to solve problems that the developer can control.

**Comment by obi1kenobi** [comment]
I agree it would be very nice to have. Unfortunately, at this we have very limited funding and resources, and none of it is currently allocated to this issue due to the substantial investment required to make any positive difference on this issue.

I'd encourage folks (especially companies!) interested in moving this forward to help us secure corporate funding to drive this feature forward. I'm currently at the limit of what I can maintain on my own, and it would be great if we could afford to hire some help.

**Comment by kornelski** [comment]
You could make builds more predictable by copying `Cargo.lock` from the local workspace to the temporary workspace used for building docs, for both current and baseline.

When you don't pass `--locked`, Cargo is quite forgiving about incomplete or mismatched lockfiles. It will take whatever is there as a starting point, and fix it to match what's being built. Crates can be missing, or be superfluous, doesn't matter. 

It will fix problems with yanked dependencies #1271 as long as the `current` workspace still has them locked.

**Comment by andreascasapu** [comment]
I don't mean to add any pressure here - it's understandable if there is limited capacity and if this feature won't make it in for some time - but I wanted to document my use case here as well, which would benefit from lockfile support:

I have a project which transitively depends on both arrow-arith and chrono. While the versions in my project's lockfile have been working together without issue, when running `cargo semver-checks --baseline-rev <ANY_EARLIER_REV>`, it fails to compile due to
```rust
 90 |         DatePart::Quarter => |d| d.quarter() as i32,
    |                                    ^^^^^^^ multiple `quarter` found
```
from both `ChronoDateExt` and `Datelike`.

Hence, I am unable to semver check the project, since, even if I narrowed/bumped the versions in the toml, older commits would still exhibit the issue. Meanwhile, the repo has had lockfiles since the beginning, so if those versions were used, the above would not be an issue (at least in my build; it's still a valuable warning that downstream users may face issues with my library, but I would still want to know if I violated semver in addition to that).
